src/main-process/atom-window.coffee

```diff
@@ -128,8 +128,12 @@ class AtomWindow
         false
 
   handleEvents: ->
-    @browserWindow.on 'close', =>
-      @atomApplication.saveState(false)
+    @browserWindow.on 'close', (event) =>
+      unless @atomApplication.quitting or @unloading
+        event.preventDefault()
+        @unloading = true
+        @atomApplication.saveState(false)
+        @saveState().then(=> @close())
```

Comment:
Should we wait to assign `@unloading = true` until after the close promise resolves? I could see a weird case where `close` is fired twice before the state finishes saving.